lib/wallet/src/Cardano/Wallet/Delegation.hs

```diff
@@ -0,0 +1,226 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.Wallet.Delegation
+    ( joinStakePoolDelegationAction
+    , guardJoin
+    , quitStakePool
+    , guardQuit
+    , quitStakePoolDelegationAction
+    , DelegationRequest(..)
+    , handleDelegationRequest
+    ) where
+
+import Prelude
+
+import qualified Cardano.Wallet.Primitive.Types as W
+import qualified Data.Set as Set
+
+import Cardano.Pool.Types
+    ( PoolId (..) )
+import Cardano.Wallet
+    ( ErrCannotQuit (..)
+    , ErrNoSuchWallet (..)
+    , ErrStakePoolDelegation (..)
+    , PoolRetirementEpochInfo (..)
+    , WalletException (..)
+    , WalletLog (..)
+    , fetchRewardBalance
+    , readRewardAccount
+    , transactionExpirySlot
+    , withNoSuchWallet
+    )
+import Cardano.Wallet.DB
+    ( DBLayer (..) )
+import Cardano.Wallet.Network
+    ( NetworkLayer (..) )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( NetworkDiscriminant )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState (..) )
+import Cardano.Wallet.Primitive.Slotting
+    ( PastHorizonException, TimeInterpreter )
+import Cardano.Wallet.Primitive.Types
+    ( Block (..)
+    , IsDelegatingTo (..)
+    , PoolLifeCycleStatus
+    , WalletDelegation (..)
+    , WalletId (..)
+    )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Transaction
+    ( DelegationAction (..)
+    , ErrCannotJoin (..)
+    , TransactionCtx
+    , Withdrawal (..)
+    , defaultTransactionCtx
+    , txDelegationAction
+    , txValidityInterval
+    , txWithdrawal
+    )
+import Control.Error
+    ( lastMay )
+import Control.Exception
+    ( throwIO )
+import Control.Monad
+    ( forM_, unless, when )
+import Control.Monad.Except
+    ( ExceptT, mapExceptT, runExceptT, withExceptT )
+import Control.Monad.IO.Class
+    ( MonadIO (..) )
+import Control.Monad.Trans.Except
+    ( except )
+import Control.Tracer
+    ( Tracer, traceWith )
+import Data.Generics.Internal.VL.Lens
+    ( view )
+import Data.Set
+    ( Set )
+
+-- | The data type that represents client's delegation request.
+-- Stake key registration is made implicit by design:
+-- the library figures out if stake key needs to be registered first
+-- so that clients don't have to worry about this concern.
+data DelegationRequest
+    = StartDelegatingRegisteringKey PoolId
+    -- ^ Delegate to a pool using the default staking key (derivation index 0),
+    -- registering the stake key if needed.
+    | StopDelegating
```

Comment:
Perhaps it would make sense to call `Join` and `Quit` for consistency with the API?

I see Join and Quit would clash with this type:
```haskell
-- | Whether the user is attempting any particular delegation action.
data DelegationAction = RegisterKeyAndJoin PoolId | Join PoolId | Quit
    deriving (Show, Eq, Generic)
```

But it seems more logical to me that `Join`/`Quit` consistently be high-level and some other name be chosen for `DelegationAction` more similar to the ledger (the lowest-level):

```haskell
data DCert
  = DCertDelegRegKey StakingCredential
  | DCertDelegDeRegKey StakingCredential
  | DCertDelegDelegate
  ...
```

Note: there is also
```haskell
data ApiDelegationAction = Join (ApiT PoolId) | Quit
    deriving (Eq, Generic, Show)
    deriving anyclass NFData
```
which clashes a bit with the name `DelegationRequest` here.

Reply:
> some other name be chosen for DelegationAction more similar to the ledger (the lowest-level):

In that case, possibly:
```haskell
-- | Whether the user is attempting any particular delegation action.
data DelegationAction = RegisterKeyAndDelegate PoolId | Delegate PoolId | DeregisterKey
    deriving (Show, Eq, Generic)
```

Reply:
Agreed.